### PR TITLE
fix `pytorchWithCuda`, fix `cupy`, upgrade `cudnn`

### LINKED ISSUE #161843: `python3Packages.pytorchWithCuda` build failure on x86_64-linux as of `3058e76a`
## Issue description
Build of `python3Packages.pytorchWithCuda` failed on x86_64-linux as of 3058e76a8925f794f64bec97e9fbea8d155fb817.

[build log](https://github.com/samuela/nixpkgs-upkeep/actions/runs/1897620867)

Maintainer cc: @teh @thoughtpolice @tscholak
Other cc: n/a

### Steps to reproduce
1. Checkout nixpkgs at commit 3058e76a8925f794f64bec97e9fbea8d155fb817
2. Run `nix-build -A python3Packages.pytorchWithCuda`

## Technical details
```
- system: `"x86_64-linux"`
 - host os: `Linux 5.11.0-1028-azure, Ubuntu, 20.04.3 LTS (Focal Fossa)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.6.1`
 - nixpkgs: `/nix/store/6v2wg6ycxzrqwn5g2pxfvx5700vddq9a-source`
```

## Misc.
This issue was automatically generated by [nixpkgs-upkeep](https://github.com/samuela/nixpkgs-upkeep).
- [CI workflow](https://github.com/samuela/nixpkgs-upkeep/actions/runs/1897620867) that created this issue.
- Internal tag: -8122276283966126424

### PR DESCRIPTION
**Pull request by samuela**
###### Description of changes
This PR accomplishes a few separate but related tasks:
* Fix pytorchWithCuda. This is concurrent work to https://github.com/NixOS/nixpkgs/pull/166734.
* Upgrade cudnn 7.6.5 -> 8.3.2.
* Fix cutensor_cudatoolkit_{10,11}.
* Harden the build of cupy, and enable parallel building.

Although I try to keep PRs as small as possible, in this case  it became clear that it was much more straightforward to make these changes simultaneously than separately. I understand that increases the burden on reviewers to some extent, but I tried to keep each commit as small as possible to ease the process.

This PR is the equivalent of https://github.com/NixOS/nixpkgs/pull/166533 but targeting master instead of staging.

According to nixpkgs-review this affects packages `pytorchWithCuda`, `Theano`, and `cupy`. I've confirmed that all of those nix-build successfully. For some reason, `nixpkgs-review wip` doesn't work in this case, but everything seems to build just fine.

Fixes https://github.com/NixOS/nixpkgs/issues/161843, https://github.com/NixOS/nixpkgs/issues/166403, and possibly others.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/science/math/cudnn/default.nix [hunk 1]
```diff
@@ -81,31 +81,27 @@ rec {
...
+      fetchurl {
+        # Starting at version 8.3.1 there's a new directory layout including
+        # a subdirectory `local_installers`.
+        url = "https://developer.download.nvidia.com/compute/redist/cudnn/v${version}/local_installers/${v}/cudnn-linux-x86_64-8.3.2.44_cuda${v}-archive.tar.xz";
+        hash = {
+          "10.2" = "sha256-1vVu+cqM+PketzIQumw9ykm6REbBZhv6/lXB7EC2aaw=";
+          "11.5" = "sha256-VQCVPAjF5dHd3P2iNPnvvdzb5DpTsm3AqCxyP6FwxFc=";
+        }."${v}";
+      };
+  };
```
Comment (samuela) [inline pkgs/development/libraries/science/math/cudnn/default.nix:104]: This change is purely a matter of formatting.

File: pkgs/development/python-modules/cupy/default.nix [hunk 2]
```diff
@@ -5,6 +5,10 @@
 , addOpenGLRunpath
 }:
 
+assert cudnn.cudatoolkit == cudatoolkit;
+assert cutensor.cudatoolkit == cudatoolkit;
+assert nccl.cudatoolkit == cudatoolkit;
```
Comment (samuela) [inline pkgs/development/python-modules/cupy/default.nix:10]: It was discovered [here](https://github.com/NixOS/nixpkgs/pull/166533#issuecomment-1084983987) that cupy was accidentally pulling in multiple cudatoolkit versions. These asserts should prevent that issue going forward.

File: pkgs/development/libraries/science/math/nccl/default.nix [hunk 3]
```diff
@@ -38,6 +38,10 @@ stdenv.mkDerivation rec {
 
   enableParallelBuilding = true;
 
+  passthru = {
+    inherit cudatoolkit;
+  };
```
Comment (samuela) [inline pkgs/development/libraries/science/math/nccl/default.nix:43]: This is necessary for the `assert` in cupy/default.nix.

**Comment by samuela** [comment]
> Breaks torchvision:

Weird it works for me:
```
❯ nix-build --max-jobs 64 -A python3Packages.torchvision     
...
copying path '/nix/store/hrwag76jyxf72fc5rk74vnqw0pl8drbf-python3.9-torchvision-0.11.3' from 'https://cache.nixos.org'...
/nix/store/hrwag76jyxf72fc5rk74vnqw0pl8drbf-python3.9-torchvision-0.11.3
```
Perhaps we're doing something slightly differently? I'm building directly off of this branch (as opposed to grafting onto latest master)

**Comment by freemagma** [comment]
Not sure if I'm doing something wrong or if this is an issue with the PR, but when I try to build pytorch with CUDA 11 on this PR I still get the error with "EmbeddingBag.cu."

```nix
{
  description = "my project description";

  inputs.nixpkgs.url = "github:NixOS/nixpkgs/2a0acdbe9f36e3dbf4e4cbcf867f2c1e990f781f";
  inputs.flake-utils.url = "github:numtide/flake-utils";

  outputs = { self, nixpkgs, flake-utils }:
    flake-utils.lib.eachDefaultSystem (system:
      with (import nixpkgs {
        config = {
          allowUnfree = true;
          cudaSupport = true;
          overlays = [
            (final: super: {
              cudatoolkit = super.cudatoolkit_11;
              cudnn = super.cudnn_8_3_cudatoolkit_11;
            })
          ];
        };
        inherit system;
      }); {
        devShell =
          let
            python = python39.override
              {
                packageOverrides = self: super: {
                  pytorch = super.pytorch.override {
                    cudatoolkit = pkgs.cudatoolkit_11;
                    cudnn = pkgs.cudnn_8_3_cudatoolkit_11;
                    magma = pkgs.magma.override { cudatoolkit = pkgs.cudatoolkit_11; };
                    nccl = pkgs.nccl.override { cudatoolkit = pkgs.cudatoolkit_11; };
                  };
                };
              };
          in
          mkShell rec {
            buildInputs = with python.pkgs; [
              pytorch
            ];
          };
      });
}
```

```
$ nix develop
...
error: builder for '/nix/store/wmf91pja41azq6mbnpfh55lvbsbscy8p-python3.9-pytorch-1.10.2.drv' failed with exit code 1;
       last 10 log lines:
       > /build/source/aten/src/ATen/native/cuda/EmbeddingBag.cu(178): here
       >
       > 41 errors detected in the compilation of "/build/source/aten/src/ATen/native/cuda/EmbeddingBag.cu".
       > CMake Error at torch_cuda_generated_EmbeddingBag.cu.o.Release.cmake:281 (message):
       >   Error generating file
       >   /build/source/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_EmbeddingBag.cu.o
       >
       >
       > [5588/6183] Building NVCC (Device) object caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_TensorModeKernel.cu.o.o[KK
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/wmf91pja41azq6mbnpfh55lvbsbscy8p-python3.9-pytorch-1.10.2.drv'.
error: 1 dependencies of derivation '/nix/store/173gfp0vxwan4ihh88dh1k8rc7n99d4z-nix-shell-env.drv' failed to build
```

**Comment by samuela** [comment]
> Not sure if I'm doing something wrong or if this is an issue with the PR, but when I try to build pytorch with CUDA 11 on this PR I still get the error with "EmbeddingBag.cu."

This is bc you're overriding pytorch with CUDA 11, which as of yet does not work. Once staging gets merged into master, we'll get the newer pytorch version, v1.11.0, which will be building with CUDA 11 by default. As to why the current pytorch version does not build successfully with CUDA 11, I have no idea unfortunately. It's very finicky to build :/

**Comment by SomeoneSerge** [comment]
> torchvision
> ... works for me
> @samuela 

You're right, the vanilla build succeeds, it's the overlay with cuda 11 that breaks.
It breaks because now calling pytorch with custom values, overriding just the unversioned top-level attributes has no effect and the torchvision build ends up consuming several versions of cuda.
This has always been a problem (e.g. with tensorflow), and there are old related problems (overriding cuda at the user site and forgetting to override cudnn) which we need to address, but don't need to address in this PR


> pytorch with CUDA 11
> @freemagma 

I can confirm that pytorch 1.10 fails to build with cuda=11.5 ([attempted hotfix with cuda 11.5](https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/129)) and succeeds to build with cuda=10.2 ([this pr](https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/121), and [a hotfix on top of unstable](https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/133), see `neverBreak.vanilla.pyPackages.pytorch`)

In my home config I'm building pytorch 1.10 from nixpkgs-21.11 with cuda 11.4 and it works, but I'm still struggling with master

**Comment by SomeoneSerge** [comment]
Build succeeds with CUDA 11.3: https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/139
Torchvision succeeds in the overlay where cuda11 is used globally

**Comment by FRidh** [comment]
> there are old related problems (overriding cuda at the user site and forgetting to override cudnn) which we need to address, but don't need to address in this PR

It might be an idea to use `passthru`. For example, having `cudnn` in the `passthru` of `cudatoolkit` and add `cudatoolkit` to the `passthru` of packages using it. Another option is to make a `cudaPackages` set for each cuda version that contains `cudnn` and others and add that in the `passthru` of the `cudatoolkit` package. Similar to what we do with say Python.

**Comment by SomeoneSerge** [comment]
> It might be an idea to use passthru
> ... another option is to make a `cudaPackages` for each cuda version

Indeed, we've been considering both, the first as a temporary fix, and the latter as a longer-term solution.
In this PR the `passthru` is used for `assert`s which should already somewhat improve the experience.
Note that just dropping `cudatoolkit` argument in e.g. `pytorch/default.nix` and using the `passthru` does not eliminate the need for `assert`s, because all of `magma`, `nccl`, and `cudnn` bring in their `cudatoolkit` and need to be consistent.
The `cudaPackages` is probably the best UX we can get at the moment

> For example, having cudnn in the passthru of cudatoolkit 

Hmm, I see. Currently it's the other way around (cudnn has cudatoolkit in passthru). Also, there are several releases of cudnn in nixpkgs. Again, `cudaPackages` seems to be a solution

**Comment by samuela** [comment]
> > there are old related problems (overriding cuda at the user site and forgetting to override cudnn) which we need to address, but don't need to address in this PR
> 
> It might be an idea to use `passthru`. For example, having `cudnn` in the `passthru` of `cudatoolkit` and add `cudatoolkit` to the `passthru` of packages using it. Another option is to make a `cudaPackages` set for each cuda version that contains `cudnn` and others and add that in the `passthru` of the `cudatoolkit` package. Similar to what we do with say Python.

On the topic of a cudaPackages package set, check out the proposal here: https://github.com/NixOS/nixpkgs/issues/163704

Re passthru: I think that's absolutely the way to go for the time being. This PR is already rather beefy, so I'd prefer to accomplish that separately, but I totally agree that we ought to do so.

**Comment by samuela** [comment]
@FRidh @SomeoneSerge Created an issue to track the torchvision cudatoolkit discrepancy: https://github.com/NixOS/nixpkgs/issues/166948

**Comment by samuela** [comment]
If no one objects, I'll go ahead and merge tomorrow

**Comment by SomeoneSerge** [comment]
`python39Packages.caffe` used to succeed and now fails with

```
caffe> -- Found cuDNN: ver. ??? found (include: /nix/store/jrj8x3764iq0iz2gqa1pdv24l08ihl6k-cudatoolkit-10.2-cudnn-8.3.2/include, library: /nix/store/
jrj8x3764iq0iz2gqa1pdv24l08ihl6k-cudatoolkit-10.2-cudnn-8.3.2/lib/libcudnn.so)
caffe> CMake Error at cmake/Cuda.cmake:221 (message):
caffe>   cuDNN version >3 is required.
```

**Comment by samuela** [comment]
> python39Packages.caffe used to succeed and now [fails](https://hercules-ci.com/accounts/github/SomeoneSerge/derivations/%2Fnix%2Fstore%2F86zrhcbawwmy41w611fm414v6ad9cks9-caffe-1.0.drv/log?via-job=6155d6f6-3611-42c3-9d11-7aff2c36e9d9) with

Hmm I'm not seeing this: 
```
ubuntu in threedoo in 172.30.0.108 nixpkgs on  samuela/cudnn3 took 16s
❯ nix-build --max-jobs 64 -A python3Packages.caffe
/nix/store/bbca44xgdj1nl19lzk70kni0hjdppagf-caffe-1.0-bin
```

**Comment by SomeoneSerge** [comment]
Relaying a conversation from [`#cuda:matrix.org`](https://matrix.to/#/#cuda:nixos.org), there's an alternative pathway.
Instead of keeping `cudatoolkit = cudatoolkit_10` and passing custom cuda, cudnn, magma and nccl to pytorch, we can do:

- Major updates for both cuda and cudnn in this PR, fixing pytorch and cupy
  
  In particular:
  - `cudatoolkit = cudatoolkit_11 = cudatoolkit_11_4`
  - `cudnn = cudnn_8_3_cudatoolkit_11`

  There's a jobset running for this combination: [143](https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/143). But I've already verified that the current pytorch (1.10) builds with cuda 11.3 and 11.4 (reminder: it has been failing with 11.5 and, apparently, 11.1; the `EmbeddingBag.cu` error)

  We revert passing additional arguments in the call to the pytorch derivation (`cudatoolkit`, ..., `nccl`), simplifying the PR and keeping the simple interface for overlays. This we would otherwise still do later, because `pytorch=1.11` wouldn't require overriding
- When pytorch 1.11 (now in staging) reaches master, we set `cudatoolkit_11 = cudatoolkit_11_5`, which is a minor update from 11.4 to 11.5 (context: the new pytorch doesn't have trouble building with cuda 11.5)

I think in this case there will be exactly one package left that still uses cuda 10 (seems to be `libtensorflow-bin`)

I'm not opposed to either solution. I prefer this alternative

**Comment by SomeoneSerge** [comment]
> Hmm I'm not seeing this:
> ❯ nix-build --max-jobs 64 -A python3Packages.caffe

Interesting. Actually, I have not rerun the job for `python310Packages.caffe` (I had this cachix issue)

Can you check `python39Packages.caffe`?

**Comment by SomeoneSerge** [comment]
Actually, just got the same error with python310. I'm wondering if I'm on the right branch 🤣

**Comment by samuela** [comment]
@SomeoneSerge I added a `caffeWithCuda` alias in https://github.com/NixOS/nixpkgs/pull/166784/commits/a60ab2a1c73453828f81d52a6a3fa3d4586c6fdf and it's building successfully now. I didn't change anything with caffe2 as that is marked as broken rn. So now everything is building successfully as far as I'm aware.

**Comment by samuela** [comment]
> Major updates for both cuda and cudnn in this PR, fixing pytorch and cupy

I'm in favor of doing this, although I hesitate to include those changes in this PR as it has gotten so huge already. What if we merged this one and then followed up immediately after with a PR that bumps the cudatoolkit and cudatoolkit_11 versions? I believe that would get us the same end result but with smaller steps along the way.

**Comment by SomeoneSerge** [comment]
I thought the major bumps would allow to drop some of the changes and reduce the size of PR... but actually, merging the asserts and fixing pytorch+cupy before cleaning up sounds tempting

File: pkgs/development/libraries/science/math/cutensor/generic.nix [hunk 4]
```diff
@@ -21,7 +21,7 @@ stdenv.mkDerivation {
 
   src = fetchurl {
     url = "https://developer.download.nvidia.com/compute/cutensor/${mostOfVersion}/local_installers/libcutensor-${stdenv.hostPlatform.parsed.kernel.name}-${stdenv.hostPlatform.parsed.cpu.name}-${version}.tar.gz";
-    inherit sha256;
+    inherit hash;
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/cutensor/generic.nix:24]: Why `hash` vs `sha256`?

File: pkgs/development/libraries/science/math/nccl/default.nix (hunk 3 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/nccl/default.nix:43]: 👍🏻

**Comment by samuela** [comment]
> I thought the major bumps would allow to drop some of the changes and reduce the size of PR... but actually, merging the asserts and fixing pytorch+cupy before cleaning up sounds tempting

Ah I see what you mean. Perhaps that is true. But I suspect that until we get pytorch v1.11 we won't be able to get out of specifying custom cudatoolkit versions since I don't think v1.10.2 supports cudatoolkit = cudatoolkit_11 = cudatoolkit_11_5. I'm not entirely sure though.

**Comment by SomeoneSerge** [comment]
>  since I don't think v1.10.2 supports cudatoolkit = cudatoolkit_11 = cudatoolkit_11_5. I'm not entirely sure though.

This is why I was talking about a downgrade to 11.4

**Comment by samuela** [comment]
Ohh I see. sorry I misunderstood!

**Review by SomeoneSerge — APPROVED** [review]
Overall, I feel OK merging this now and immediately following up with a clean-up PR that would adjust the cudnn and cudatoolkit attributes so as to minimize the total number of ad hoc parameters passed to `callPackage`s: the first step makes the attributes of nixpkgs buildable, the second fixes the overlays UX, which seems like a reasonable allocation of priorities

File: pkgs/development/libraries/science/math/cutensor/generic.nix (hunk 4 above)
Comment (samuela) [reply pkgs/development/libraries/science/math/cutensor/generic.nix:24]: Iiuc hash is preferred to sha256 as it is more future proof, but I don't remember the reference for that

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/cutensor/generic.nix:24]:
```
, # SRI hash.
  hash ? ""

, # Legacy ways of specifying the hash.
  outputHash ? ""
, outputHashAlgo ? ""
, md5 ? ""
, sha1 ? ""
, sha256 ? ""
, sha512 ? ""
```

Hmm, I guess

Comment (FRidh) [reply pkgs/development/libraries/science/math/cutensor/generic.nix:24]: yep, we are switching everywhere (though slowly) to SRI hashes.

**Pull request merged**